src/toxcore/dht/codec.rs

```diff
@@ -33,23 +90,90 @@ pub enum DecodeError {
     },
     /// Error indicates that received packet can't be parsed
     #[fail(display = "Deserialize Packet error: {:?}, packet: {:?}", error, packet)]
-    DeserializeError {
+    Deserialize {
         /// Parsing error
         error: ErrorKind,
         /// Received packet
         packet: Vec<u8>,
+    },
+    /// For From<DecodeError> trait.
+    #[fail(display = "IO Error")]
+    Io,
+}
+
+/// Error that can happen when encoding `Packet` to bytes
+#[derive(Debug)]
+pub struct EncodeError {
+    ctx: Context<EncodeErrorKind>,
+}
+
+impl EncodeError {
+    /// Return the kind of this error.
+    pub fn kind(&self) -> &EncodeErrorKind {
+        self.ctx.get_context()
+    }
+
+    pub(crate) fn serialize(error: GenError) -> EncodeError {
+        EncodeError::from(EncodeErrorKind::Serialize { error })
+    }
+}
+
+impl Fail for EncodeError {
+    fn cause(&self) -> Option<&Fail> {
+        self.ctx.cause()
+    }
+
+    fn backtrace(&self) -> Option<&Backtrace> {
+        self.ctx.backtrace()
+    }
+}
+
+impl fmt::Display for EncodeError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        self.ctx.fmt(f)
+    }
+}
+
+impl From<EncodeErrorKind> for EncodeError {
+    fn from(kind: EncodeErrorKind) -> EncodeError {
+        EncodeError::from(Context::new(kind))
+    }
+}
+
+impl From<Context<EncodeErrorKind>> for EncodeError {
+    fn from(ctx: Context<EncodeErrorKind>) -> EncodeError {
+        EncodeError { ctx }
+    }
+}
+
+impl From<IoError> for DecodeError {
+    fn from(error: IoError) -> DecodeError {
+        DecodeError {
+            ctx: error.context(DecodeErrorKind::Io)
+        }
+    }
+}
+
+impl From<IoError> for EncodeError {
+    fn from(error: IoError) -> EncodeError {
+        EncodeError {
+            ctx: error.context(EncodeErrorKind::Io)
+        }
     }
 }
 
 /// Error that can happen when encoding `Packet` to bytes
-#[derive(Debug, Fail)]
-pub enum EncodeError {
+#[derive(Debug, PartialEq, Fail)]
+pub enum EncodeErrorKind {
     /// Error indicates that `Packet` is invalid and can't be serialized
     #[fail(display = "Serialize Packet error: {:?}", error)]
-    SerializeError {
+    Serialize {
         /// Serialization error
         error: GenError
-    }
+    },
+    /// For From<EncodeError> trait.
```

Comment:
It's for From<IoError> for EncodeError, but I'd write something like:
```rust
/// General IO error that can happen with UDP socket.
```
Otherwise looks good.

Reply:
done